crates/host/src/nats/policy.rs

```diff
@@ -24,14 +25,119 @@ use crate::policy::{
     POLICY_TYPE_VERSION,
 };
 
+#[derive(Debug)]
+struct CacheEntry<T>
+where
+    T: Clone,
+{
+    value: T,
+    last_updated: Instant,
+}
+
+impl<T> CacheEntry<T>
+where
+    T: Clone,
+{
+    fn new(value: T) -> Self {
+        Self {
+            value,
+            last_updated: Instant::now(),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+struct DecisionCache<K, V>
+where
+    K: std::hash::Hash + Eq + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    cache: Arc<RwLock<HashMap<K, CacheEntry<V>>>>,
+    ttl: Duration,
+    _cleanup_handle: tokio::task::AbortHandle,
+}
+
+impl<K, V> DecisionCache<K, V>
+where
+    K: std::hash::Hash + Eq + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn new(ttl: Duration) -> Self {
+        Self::with_cleanup(ttl, ttl / 2)
+    }
+
+    fn with_cleanup(ttl: Duration, interval: Duration) -> Self {
+        let cache = Arc::new(RwLock::new(HashMap::<K, CacheEntry<V>>::new()));
+        let cleanup_cache = Arc::clone(&cache);
+        let cleanup_handle = spawn(async move {
```

Comment:
This is going to be pretty write heavy which could cause contention, especially under a large amount of load since it scans _all_ cached responses every time it ticks. I think there are two approaches here. If you want to roll it by hand, you can combine "lazy" expiration (checking the TTL on a `get` and removing if expired) with a cleanup loop that runs less often (once every few minutes) to clean up expired entries.

The easier option might be to use the [`DelayQueue`](https://docs.rs/tokio-util/latest/tokio_util/time/struct.DelayQueue.html) from tokio_util. It has an example in there of a simple TTL cache there that is fairly helpful.

Reply:
I followed your recommendation and used the DelayQueue. I also read a bit, and it seems that it tends to be more performant to do it based on cycles rather than time. So, I still expect entry based on time, but I only triggered the purge after a certain number of write cycles. lmk your thoughts